File: src/data/courses/dsa/binary-search-on-answer/capacity-time-optimization/minimum-days-bouquets/code/java/MinimumDaysBouquets.java
class Solution {
    public int minDays(int[] bloomDay, int m, int k) {
        if ((long)m*k > bloomDay.length) return -1;
        int lo=1, hi=0; for(int d:bloomDay) hi=Math.max(hi,d);
        while(lo<hi) { int mid=lo+(hi-lo)/2,b=0,f=0; for(int d:bloomDay){if(d<=mid){if(++f==k){b++;f=0;}}else f=0;} if(b>=m)hi=mid;else lo=mid+1; }
        return lo;
    }
}

File: src/data/courses/dsa/binary-search-on-answer/capacity-time-optimization/minimum-days-bouquets/code/python/minimum_days_bouquets.py
class Solution:
    def minDays(self, bloomDay: List[int], m: int, k: int) -> int:
        if m * k > len(bloomDay): return -1
        lo, hi = 1, max(bloomDay)
        while lo < hi:
            mid = (lo + hi) // 2
            bouquets = flowers = 0
            for d in bloomDay:
                if d <= mid: flowers += 1
                else: flowers = 0
                if flowers == k: bouquets += 1; flowers = 0
            if bouquets >= m: hi = mid
            else: lo = mid + 1
        return lo